### PR TITLE
record: ensure block.written is accessed atomically

### LINKED ISSUE #1692: record: non atomic access of atomic field
This was noticed in cockroachdb/cockroach#81511.

This field is declared to be atomic:

https://github.com/cockroachdb/pebble/blob/9bd431430a38da6564dc46e2adfcf8ff77e10c23/record/log_writer.go#L25-L26

Yet we read from it non-atomically here:

https://github.com/cockroachdb/pebble/blob/9bd431430a38da6564dc46e2adfcf8ff77e10c23/record/log_writer.go#L678

A trivial fix would be to update the read-site to read atomically.

While we're in here, we should audit the read/write call-sites to make sure we're honoring the atomic call-site.

It may even be worth some sugar in the struct to highlight the fact that the field is atomic (i.e. `block.atomic.writer`). I believe I've seen this patten adopted elsewhere.

### PR DESCRIPTION
**Pull request by nicktrav**
Currently, a block's `written` field is intended to be used atomically.
There are existing call sites where the field is read from or written to
non-atomically, which can result in a data race (as observed in
cockroachdb/cockroach#81511).

Make reads and writes of `block.written` atomic.

This results in up to a 8% reduction in latency and increase in
throughput, depending on the size of the record, due to a change to use
a stack variable rather that dereferencing a field on the struct.
However, it should be noted that this particular benchmark does not
touch the filesystem (it makes use of `io.Discard`). In practice the
savings would be dwarfed by the filesystem operation (sync / write).

```
name                       old time/op    new time/op    delta
RecordWrite/size=8-16        42.9ns ± 5%    40.7ns ± 2%  -5.24%  (p=0.000 n=10+10)
RecordWrite/size=16-16       42.7ns ± 4%    40.7ns ± 1%  -4.89%  (p=0.000 n=10+10)
RecordWrite/size=32-16       44.0ns ± 6%    40.4ns ± 1%  -8.24%  (p=0.000 n=10+10)
RecordWrite/size=64-16       44.1ns ± 1%    41.6ns ± 0%  -5.58%  (p=0.000 n=10+10)
RecordWrite/size=256-16      67.7ns ± 4%    65.2ns ± 1%  -3.71%  (p=0.000 n=10+9)
RecordWrite/size=1028-16      124ns ± 2%     119ns ± 1%  -4.04%  (p=0.000 n=10+10)
RecordWrite/size=4096-16      306ns ± 5%     296ns ± 1%  -3.25%  (p=0.000 n=10+9)
RecordWrite/size=65536-16    4.54µs ± 1%    4.52µs ± 2%    ~     (p=0.247 n=10+10)

name                       old speed      new speed      delta
RecordWrite/size=8-16       186MB/s ± 5%   197MB/s ± 2%  +5.49%  (p=0.000 n=10+10)
RecordWrite/size=16-16      374MB/s ± 3%   394MB/s ± 1%  +5.12%  (p=0.000 n=10+10)
RecordWrite/size=32-16      727MB/s ± 5%   792MB/s ± 1%  +8.91%  (p=0.000 n=10+10)
RecordWrite/size=64-16     1.45GB/s ± 1%  1.54GB/s ± 0%  +5.91%  (p=0.000 n=10+10)
RecordWrite/size=256-16    3.78GB/s ± 4%  3.92GB/s ± 1%  +3.82%  (p=0.000 n=10+9)
RecordWrite/size=1028-16   8.31GB/s ± 2%  8.66GB/s ± 1%  +4.21%  (p=0.000 n=10+10)
RecordWrite/size=4096-16   13.4GB/s ± 5%  13.8GB/s ± 1%  +3.30%  (p=0.000 n=10+9)
RecordWrite/size=65536-16  14.4GB/s ± 1%  14.5GB/s ± 2%    ~     (p=0.247 n=10+10)

name                       old alloc/op   new alloc/op   delta
RecordWrite/size=8-16         0.00B          0.00B         ~     (all equal)
RecordWrite/size=16-16        0.00B          0.00B         ~     (all equal)
RecordWrite/size=32-16        0.00B          0.00B         ~     (all equal)
RecordWrite/size=64-16        0.00B          0.00B         ~     (all equal)
RecordWrite/size=256-16       0.00B          0.00B         ~     (all equal)
RecordWrite/size=1028-16      0.00B          0.00B         ~     (all equal)
RecordWrite/size=4096-16      0.00B          0.00B         ~     (all equal)
RecordWrite/size=65536-16     2.00B ± 0%     2.00B ± 0%    ~     (all equal)

name                       old allocs/op  new allocs/op  delta
RecordWrite/size=8-16          0.00           0.00         ~     (all equal)
RecordWrite/size=16-16         0.00           0.00         ~     (all equal)
RecordWrite/size=32-16         0.00           0.00         ~     (all equal)
RecordWrite/size=64-16         0.00           0.00         ~     (all equal)
RecordWrite/size=256-16        0.00           0.00         ~     (all equal)
RecordWrite/size=1028-16       0.00           0.00         ~     (all equal)
RecordWrite/size=4096-16       0.00           0.00         ~     (all equal)
RecordWrite/size=65536-16      0.00           0.00         ~     (all equal)
```

Fixes #1692.

### CODE REVIEW

**Comment by nicktrav** [comment]
This is just one approach to solving this issue - honoring the contract that the struct field comment implies re: atomicity.

It looks like there are also some comments around using the commit pipeline mutex for external synchronization ([for example](https://github.com/cockroachdb/pebble/blob/bb2c1501ac23ca3ce17d46717986610424b05790/record/log_writer.go#L582)), though evidently something is slipping through the cracks are we're seeing the race in Cockroach.

For context, in cockroachdb/cockroach#81133, parallel writes are made to each engine's WAL as part of liveness checking, which can race if the context is cancelled and the log is closed.

Seems like this code in Pebble has been this way since inception, and we're only tickling the issue now due to the nature of the concurrent access introduced in the Cockroach PR.

**Review by nicktrav — COMMENTED** [review]
> I believe its synchronization is designed for a concurrent single-consumer, single-producer scheme

This was my read too. The commentary was a little confusing though.

While this patch fixes this particular issue, we're still getting a race (when used in Cockroach) between the following two call sites when we close the writer:

Read: 
https://github.com/cockroachdb/pebble/blob/bb2c1501ac23ca3ce17d46717986610424b05790/record/log_writer.go#L677

Write:

https://github.com/cockroachdb/pebble/blob/bb2c1501ac23ca3ce17d46717986610424b05790/record/log_writer.go#L573

Rather than sprinkling mutexes and atomics around the place in Pebble to cover use-cases above Pebble, I wonder if there's a way we can continue to honor the existing contract and adapt what we're doing above Pebble:

> I think we should find a way to avoid the batch application racing with the closing of the database above Pebble,

Basically +1 on this part.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1719#-:-N2qXwPL94zxXbMqB-NK:br4h9ut)* status: 0 of 1 files reviewed, all discussions resolved (waiting on @sumeerbhola)

**Review by nicktrav — COMMENTED** [review]
I'm going to close this out as a signal that the code shouldn't be reviewed, but we can continue the conversation here.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1719#-:-N2qYrvH7APvagoNP99C:b-zfovj7)* status: 0 of 1 files reviewed, all discussions resolved (waiting on @sumeerbhola)

**Comment by petermattis** [comment]
`LogWriter.{Write,SyncRecord,Close}` need external synchronization. Inside Pebble this synchronization is provided by `commitPipeline.mu`. It looks like we're not holding `commitPipeline.mu` correctly in the `DB.Close()` path.

**Review by erikgrinaker — COMMENTED** [review]
> Rather than sprinkling mutexes and atomics around the place in Pebble to cover use-cases above Pebble, I wonder if there's a way we can continue to honor the existing contract and adapt what we're doing above Pebble:
> 
>> I think we should find a way to avoid the batch application racing with the closing of the database above Pebble,
> 
> Basically +1 on this part.

Ok, I'll add some synchronization in Pebble.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1719#-:-N2u8i8z0JDA1JQ1dQhr:b-l90ug7)* status: 0 of 1 files reviewed, all discussions resolved

**Review by erikgrinaker — COMMENTED** [review]
https://github.com/cockroachdb/cockroach/pull/81813

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1719#-:-N2uzdTDDg-gjz-h9osu:b-ycbfy)* status: 0 of 1 files reviewed, all discussions resolved
